tests/test_routes.py

```diff
@@ -147,3 +149,106 @@ def test_list_shopcart_items_none_items_error(self, mock_shopcart_get_by_id):
+        shopcart = self._create_an_empty_shopcart(1)[0]
+        res = self.client.post(
+            f"{BASE_URL}/{shopcart.id}/items",
+            json={},  # Empty request body
+            content_type="application/json",
+        )
+        self.assertEqual(res.status_code, status.HTTP_400_BAD_REQUEST)
+    
+    def test_add_items_with_invalid_data_type(self):
+        """ It should not be added with invalid data type """
+        shopcart = self._create_an_empty_shopcart(1)[0]
+        item = ItemFactory()
+        res = self.client.post(
+            f"{BASE_URL}/{shopcart.id}/items",
+            json=item.serialize(),
+            content_type="text/json",
+        )
+        self.assertEqual(res.status_code, status.HTTP_415_UNSUPPORTED_MEDIA_TYPE)
+
+    def test_add_items_with_invalid_quantity(self):
+        """ It should not be added with invalid quantity """
+        shopcart = self._create_an_empty_shopcart(1)[0]
+        item = ItemFactory()
+        item.quantity = 0
+        res = self.client.post(
+                f"{BASE_URL}/{shopcart.id}/items",
+                json=item.serialize(),
+                content_type="application/json",
+                )
+        self.assertEqual(res.status_code, status.HTTP_400_BAD_REQUEST)
+
+    @patch.object(Shopcart, 'get_by_id', MagicMock(side_effect=Exception("DBAPIErr")))
+    def test_add_shopcart_items_shopcart_by_id_error(self):
+        """ It should get internal server error when add items if there's exception in Shopcart.get_by_id """
+        shopcart = self._create_an_empty_shopcart(1)[0]
+        item = ItemFactory()
+        res = self.client.post(
+                f"{BASE_URL}/{shopcart.id}/items",
+                json=item.serialize(),
+                content_type="application/json",
+                )
+        self.assertEqual(res.status_code, status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+    @patch('service.models.Shopcart.get_by_id')
+    def test_add_shopcart_items_none_items_error(self, mock_shopcart_get_by_id):
+        """ It should get internal server error when add items if Shopcart.items is None """
+        mock_shopcart = Mock(items=None)
+        mock_shopcart_get_by_id.return_value = mock_shopcart
+        item = ItemFactory()
+        res = self.client.post(
+                f"{BASE_URL}/0/items",
+                json=item.serialize(),
+                content_type="application/json",
+                )
+        self.assertEqual(res.status_code, status.HTTP_500_INTERNAL_SERVER_ERROR)
+    
+    @patch.object(Shopcart, 'update')
+    def test_add_shopcart_items_data_validation_error(self, mock_shopcart_update):
+        """It should return a 400 Bad Request response for DataValidationError"""
+        shopcart = self._create_an_empty_shopcart(1)[0]
+        item = ItemFactory()
+
+        # Raise a DataValidationError when calling shopcart.update
+        mock_shopcart_update.side_effect = DataValidationError("Invalid data")
```

Comment:
nice to learn there's such syntax!

Reply:
One reminder is that when selecting type of error to be raised as side effect, we would prefer choosing one that is possible to be triggered with the statement. For `shopcart.update()`, given that it is a method implemented in the SQLAlchemy package as one of the DBAPI, we should raise a possible error type according to their source code.

I checked SQLAlchemy's source code in Github and found that [SQLAlchemyError](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/exc.py#L10) is their base exception class lol

Reply:
I think the correct way here is to import the possible error class defined in SQLAlchemyError (e.g. the DBAPIError), and set it as the side_effect. However, I haven't succeeded in figuring out how to import the DBAPIError class, that's why I chose Exception("DBAPIError") as an workaround in other test cases 😂

Reply:
Thanks for your comments!

Reply:
ok, maybe let's leave it as is until we figure out a better way, thanks~